apiserver/facades/client/application/application.go

```diff
@@ -2024,20 +2035,32 @@ func (api *APIBase) setApplicationConfig(arg params.ApplicationConfigSet) error
 	if len(charmConfig) > 0 {
 		ch, _, err := app.Charm()
 		if err != nil {
-			return err
+			return errors.Trace(err)
 		}
 		// Validate the charm and application config.
 		charmConfigChanges, err := ch.Config().ParseSettingsStrings(charmConfig)
 		if err != nil {
-			return err
+			return errors.Trace(err)
 		}
 		if err := app.UpdateCharmConfig(arg.Generation, charmConfigChanges); err != nil {
 			return errors.Annotate(err, "updating application charm settings")
 		}
+		if err := api.addAppToGeneration(arg.ApplicationName); err != nil {
```

Comment:
Why not check arg.Generation before doing this?

Reply:
Indeed. We have to. Will fix.